Command palette

Twixes opened the pull request:
## Changes

This adds a command palette, shown with Ctrl/Cmd + K, allowing for powerful keyboard-based navigation and control over PostHog UI.

## Checklist

- [ ] Backend tests (if this PR affects the backend)
- [ ] Cypress E2E tests (if this PR affects the front and/or backend)

EDsCODE committed:
add hotkey base

EDsCODE committed:
add platform check

Twixes committed:
Rename CommandBox to CommandPalette

Twixes committed:
Add styled-components

Twixes committed:
Merge branch 'command-palette' of https://github.com/PostHog/posthog into command-palette

Twixes committed:
Restructure CommandPalette and add Esc handling

Twixes committed:
Update utils.js

EDsCODE committed:
add search box

Twixes committed:
Add logic types

Twixes committed:
Implement base logic

Twixes committed:
Fix scroll prevention

Twixes committed:
Update logic and add goto command base with Fuse fuzzy serach

EDsCODE committed:
ux hotkey

EDsCODE committed:
ux hotkey

Twixes committed:
Implement command (de)registration

Twixes committed:
Merge branch 'command-palette' of https://github.com/PostHog/posthog into command-palette

EDsCODE committed:
minor fix

EDsCODE committed:
Merge branch 'command-palette' of https://github.com/PostHog/posthog into command-palette

paolodamico committed:
base UI

paolodamico committed:
Merge branch 'command-palette' of github.com:PostHog/posthog into command-palette

paolodamico committed:
error message & hints

paolodamico committed:
command groups UI

Twixes committed:
Add commands search

Twixes committed:
Merge branch 'command-palette' of https://github.com/PostHog/posthog into command-palette

Twixes committed:
Make commands work

Twixes committed:
Prepare palette for fuzzy search and add basic labeling

paolodamico committed:
ui adjustments

paolodamico committed:
ui

Twixes committed:
Improve colors

paolodamico committed:
switch to global command

paolodamico committed:
Merge branch 'command-palette' of github.com:PostHog/posthog into command-palette

Twixes committed:
Optimize resolution

Twixes committed:
Merge branch 'command-palette' of https://github.com/PostHog/posthog into command-palette

paolodamico committed:
global go to commands

paolodamico committed:
abstract item selection logic to command palette

paolodamico committed:
BM

Twixes committed:
Update casing and types

Twixes committed:
Add conditional useCommands

Twixes committed:
Improve results

EDsCODE committed:
keyboard nav

EDsCODE committed:
keyboard nav fix

paolodamico committed:
removed double fuse (synonyms broken)

Twixes committed:
Update style

Twixes committed:
Merge branch 'command-palette' of https://github.com/PostHog/posthog into command-palette

paolodamico committed:
Merge branch 'command-palette' of github.com:PostHog/posthog into command-palette

paolodamico committed:
Merge branch 'command-palette' of github.com:PostHog/posthog into command-palette

Twixes committed:
Fix highlighting

Twixes committed:
Remove extra style

paolodamico committed:
conditional go to commands

EDsCODE committed:
add mouse hover handling and rename functions

EDsCODE committed:
remove mac check

paolodamico committed:
Merge branch 'command-palette' of github.com:PostHog/posthog into command-palette

paolodamico committed:
add more commands

paolodamico committed:
added insights stub pages

Twixes committed:
Add command components

Twixes committed:
Fix things

EDsCODE committed:
refactor logic

Twixes committed:
Update styling

Twixes committed:
Merge branch 'command-palette' of https://github.com/PostHog/posthog into command-palette

Twixes committed:
Remove extraneous `input`

paolodamico committed:
load custom dashboards to command palette

paolodamico committed:
Merge branch 'command-palette' of github.com:PostHog/posthog into command-palette

Twixes committed:
Add input styling

Twixes committed:
Merge branch 'command-palette' of https://github.com/PostHog/posthog into command-palette

Twixes committed:
Add write icon

Twixes committed:
Refactor results

paolodamico committed:
minor adjustments

EDsCODE committed:
add person search

EDsCODE committed:
add person search

paolodamico committed:
remove papercups fully

EDsCODE committed:
fix input indexing

EDsCODE committed:
fix result executing empty

Twixes committed:
Add command grouping and improve navigation

Twixes committed:
Merge branch 'command-palette' of https://github.com/PostHog/posthog into command-palette

EDsCODE committed:
add urls

EDsCODE committed:
Merge branch 'command-palette' of https://github.com/PostHog/posthog into command-palette

EDsCODE committed:
make logic explicit

EDsCODE committed:
merge master

EDsCODE committed:
fix error with hadnler

Twixes committed:
Remove redundant border-top

EDsCODE committed:
add trend functions

paolodamico committed:
Merge branch 'command-palette' of github.com:PostHog/posthog into command-palette

paolodamico committed:
personal api key stub

Twixes committed:
Add command palette toggle button

Twixes committed:
Add palette toggle

Twixes committed:
Fix things

paolodamico committed:
fix personal api keys

paolodamico committed:
Merge branch 'command-palette' of github.com:PostHog/posthog into command-palette

paolodamico committed:
copy api key to clipboard

Twixes committed:
Add toggle text

Twixes committed:
Merge branch 'command-palette' of https://github.com/PostHog/posthog into command-palette

Twixes committed:
Improve palette UX and perform refactoring

Twixes committed:
Improve UX and add palette usage reporting

paolodamico committed:
tests refactor

paolodamico committed:
filter person list

paolodamico committed:
refactor api person tests

EDsCODE committed:
add squeak

EDsCODE committed:
merge master

paolodamico committed:
deprecate by_distinct_id & by_email person endpoints

Twixes committed:
Optimize squeak

EDsCODE committed:
fix typegen error

EDsCODE committed:
Merge branch 'command-palette' of https://github.com/PostHog/posthog into command-palette

paolodamico committed:
use new filters in frontend

paolodamico committed:
Merge branch 'command-palette' of github.com:PostHog/posthog into command-palette

Twixes committed:
Optimize squeak

Twixes committed:
Merge branch 'command-palette' of https://github.com/PostHog/posthog into command-palette

paolodamico committed:
key identifier refactor

paolodamico committed:
Merge branch 'command-palette' of github.com:PostHog/posthog into command-palette

paolodamico committed:
fix mypy

paolodamico committed:
removed unused code

Twixes committed:
Make custom command UI more coherent

mariusandra commented:
UX feedback without looking at the code

- This doesn't work. 

![2020-10-11 10 20 01](https://user-images.githubusercontent.com/53387/95673707-7ed2f600-0bab-11eb-8ba0-eba219e4fb5c.gif)

- It should clear it when closing the modal

![2020-10-11 10 20 27](https://user-images.githubusercontent.com/53387/95673711-84304080-0bab-11eb-9fac-82cac605b2e2.gif)

- What "suggestions"?

![image](https://user-images.githubusercontent.com/53387/95673746-cfe2ea00-0bab-11eb-9c64-bb1f815e73a7.png)

On my android TV setup, the search box keeps showing actual suggestions like "Try dashboards" ... and a few seconds later "Try searching for Katy Perry songs", etc

- Clicking up/down shouldn't move the cursor

![2020-10-11 10 24 41](https://user-images.githubusercontent.com/53387/95673762-09b3f080-0bac-11eb-81b9-2a3131135187.gif)

Clicking up could move to the last item as well

- No built in calculator

![2020-10-11 10 25 44](https://user-images.githubusercontent.com/53387/95673783-2a7c4600-0bac-11eb-841d-ace49c1b9b66.gif)

- Remove debug stuff

![image](https://user-images.githubusercontent.com/53387/95673791-3962f880-0bac-11eb-8eb8-5160fc227d37.png)

- Not responsive and the CMD+K part goes on 3 lines, should just say ⌘K or something

![2020-10-11 10 27 09](https://user-images.githubusercontent.com/53387/95673814-60212f00-0bac-11eb-93d6-ed194582a092.gif)

- Should scroll

![2020-10-11 10 28 57](https://user-images.githubusercontent.com/53387/95673871-a8405180-0bac-11eb-87e1-ff47ef5889c4.gif)

mariusandra commented on frontend/src/scenes/users/People.js:
```diff
@@ -89,7 +89,6 @@ function _People() {
             <input
                 className="form-control"
                 name="search"
-                autoFocus
```
Nice! This was annoying when using the page on mobile

mariusandra commented on posthog/api/person.py:
```diff
@@ -41,11 +42,25 @@ class CursorPagination(BaseCursorPagination):
     page_size = 100
 
 
+class PersonFilter(filters.FilterSet):
+    email = filters.CharFilter(field_name="properties__email")
+    distinct_id = filters.CharFilter(field_name="persondistinctid__distinct_id")
+    key_identifier = filters.CharFilter(method="key_identifier_filter")
+
+    def key_identifier_filter(self, queryset, attr, *args, **kwargs):
+        """
+        Filters persons by email or distinct ID
+        """
+        return queryset.filter(Q(persondistinctid__distinct_id=args[0]) | Q(properties__email=args[0]))
```
I'm not confident enough in my django to just approve this, so I'll ask: does this scope by team as well?

paolodamico replied:
This uses django-filter which **only adds extra filtering** to the existing queryset, as we're only further restricting the queryset, there should be no issues related to the team scoping, but I'll definitely add a test to be abs sure.

mariusandra commented on posthog/api/test/test_person.py:
```diff
@@ -88,44 +97,89 @@ def test_filter_by_cohort(self):
+        response = self.client.get(f"/api/person/?cohort={cohort.pk}")
+        self.assertEqual(len(response.json()["results"]), 1, response)
+
+    def test_filter_person_list(self):
+
+        person1: Person = Person.objects.create(
+            team=self.team, distinct_ids=["distinct_id", "another_one"], properties={"email": "someone@gmail.com"},
+        )
+        person2: Person = Person.objects.create(
+            team=self.team, distinct_ids=["distinct_id_2"], properties={"email": "another@gmail.com"},
+        )
+
```
would be great to add a test to explicitly make sure we don't see PII from other teams

paolodamico replied:
Definitely! I'll add this before the next review

paolodamico replied:
Ready for review

mariusandra commented on posthog/templates/overlays.html:
```diff
@@ -21,30 +21,4 @@
     
-{% endif %}
-
-{% if debug or request.get_host == 'app.posthog.com' %}
```
might be wise to put this back in

Twixes replied:
Ah well, there's an issue with this thing (Papercups chat window), in that it steals focus in a way that prevents Command + K from working before a user clicks anywhere on the page. So we'll need Papercups to fix this in their product.

mariusandra replied:
Hmm... that's a bit unfortunate. I don't think disabling papercups is the answer since we rely on it a lot now. Did you check with them if this can be fixed? Any workaround we could apply?

mariusandra reviewed (changes requested):
Great overall, some things to improve :)

Twixes committed:
Merge branch 'command-palette' of https://github.com/PostHog/posthog into command-palette

Twixes commented on frontend/src/scenes/insights/insightCommandLogic.ts:
```diff
@@ -0,0 +1,55 @@
+                        compareFilterLogic.actions.setCompare(!compareFilterLogic.values.compare)
+                    },
+                },
+                ...Object.entries(dateMapping).map(([key, value]) => ({
+                    key: `insight-${key}`,
+                    icon: RiseOutlined,
+                    display: `Set time range to ${key}`,
+                    executor: () => {
+                        dateFilterLogic.actions.setDates(value[0], value[1])
+                    },
+                })),
```
Definitely doable.

mariusandra replied:
not done? not needed tbh...

Twixes commented on frontend/src/scenes/insights/InsightTabs/TrendTab/trendCommandLogic.ts:
```diff
@@ -0,0 +1,11 @@
+import { commandLogicType } from 'types/lib/components/CommandPalette/commandLogicType'
+import { kea } from 'kea'
+
+export const trendCommandLogic = kea<commandLogicType<Command, CommandRegistrations>>({
+    connect: [commandLogic],
+    events: () => ({
+        afterMount: () => {},
+        beforeUnmount: () => {},
+    }),
+})
```
Paging @EDsCODE

Twixes commented on frontend/src/lib/components/CommandPalette/CustomCommands/ApiKeyCommand.tsx:
```diff
@@ -0,0 +1,71 @@
+    handleCancel: () => void
+}
+
+export function ApiKeyCommand({ handleCancel }: ApiKeyCommandProps): JSX.Element {
```
True. Reworking custom commands for slightly better UX and way better maintainability (currently this is mostly hard-coded) with a proper Flow Command™️ protocol.

Twixes commented on frontend/src/lib/components/CommandPalette/CustomCommands/ApiKeyCommand.tsx:
```diff
@@ -0,0 +1,71 @@
+import { apiKeyCommandLogic } from './apiKeyCommandLogic'
+import { CustomCommandBox, CommandTitle, CommandInputContainer, CommandInputElement } from '../commandStyledComponents'
+import { personalAPIKeysLogic } from 'lib/components/PersonalAPIKeys/personalAPIKeysLogic'
+import styled from 'styled-components'
```
Eh, IMO styled-components are more maintainable than `.(s)css` files (hard to leave unused styles behind, easy to keep track), but definitely agree that class names are not elegant this way and we also have to repeat some styling needlessly. @paolodamico What do you think? I'm leaning towards `.scss`.

Twixes replied:
Went with @mariusandra's suggested approach – moved styling to `.scss`.

Twixes commented on frontend/src/lib/components/CommandPalette/CommandResults.tsx:
```diff
@@ -0,0 +1,153 @@
+
+    useEventListener('keydown', handleKeyDown)
+
+    const groupedResults: { [scope: string]: CommandResultType[] } = {}
+    for (const result of commandSearchResults) {
+        const scope: string = result.command.scope
+        if (!(scope in groupedResults)) groupedResults[scope] = []
+        groupedResults[scope].push(result)
+    }
+    // Always put global commands group last
+    const sortedGroups = Object.entries(groupedResults)
+    let rollingIndex = 0
+    for (const [, group] of sortedGroups) {
+        for (const result of group) {
+            result.index = rollingIndex++
+        }
+    }
```
Good points, will address.

Twixes commented on frontend/src/lib/components/CommandPalette/CommandInput.tsx:
```diff
@@ -0,0 +1,56 @@
+import { useValues, useActions } from 'kea'
+import { commandLogic } from './commandLogic'
+import { useEventListener } from 'lib/hooks/useEventListener'
+import squeakFile from './../../../../public/squeak.mp3'
```
Squeak from hell.

Twixes replied:
It's now the PostHog Productions squeak. Still not a hedgehog – apparently it's an INSECT CRY – but it's less hellish.

Twixes commented on frontend/src/layout/TopContent/index.tsx:
```diff
@@ -0,0 +1,41 @@
+                    fontSize: 13,
+                }}
+            >
+                {!isMobile() && <CommandPaletteButton />}
```
The button won't show up, since it's essentially impossible to detect such details. But the keyboard shortcut will work regardless of that. The button is just a suggestion.

Twixes commented:
Addressing your UX points @mariusandra:
1. The Squeak is not pleasant, but it works reliably for me…
2. The input box is not cleared on palette close on purpose, this mimics Spotlight's behavior (which is pretty useful IMO).
3. Definitely need to add some command suggestions for discoverability… Should be simple enough.
4. I don't quite understand "Clicking up/down shouldn't move the cursor", could you elaborate?
5. Adding a calculator right now.
6. Will remove debugging. 👍 
7. Can hide this with a media query or narrow screens, though it shouldn't show up on mobile devices anyway (through a user agent check).
8. By "scroll", you mean show more than 5 results?

Twixes committed:
Add calculator to palette

Twixes committed:
Use equal sign

Twixes commented:
Now calculates. @mariusandra 
<img width="589" alt="Calculator" src="https://user-images.githubusercontent.com/4550621/95689697-d48eb900-0c12-11eb-97f5-8c174de2b807.png">

Twixes committed:
Make palette button nicer

Twixes committed:
Add lodash back officially and show palette suggestions

lodash is still used in a few places, but it was not in package.json. The reason this was working was that lodash is a dependency of some other depedencies, but this was fragile. It's still not ideal to use this, but at least this is now not a hack the way it was.

Twixes committed:
Remove isHint

Twixes committed:
Optimize graph time range command

Twixes committed:
Move command results grouping to Kea logic

Twixes committed:
Fix result focus autoshift

Twixes committed:
Improve palette result focusing

Twixes committed:
Adjust for window.posthog being optional

paolodamico committed:
add test for third-party person filters

paolodamico commented on frontend/src/lib/components/CommandPalette/commandStyledComponents.ts:
```diff
@@ -78,12 +77,6 @@ export const ResultBox = styled.div<ResultBoxProps>`
             width: 0.375rem;
         }
         `}
-    ${({ isHint }) =>
```
We definitely need to have hints to increase commands discoverability. This will be particularly important for commands that you don't initially expect (like share feedback). Besides we already have this code

Twixes replied:
Now we have randomized context-dependent suggestions, which are actionable as opposed to hints. An improvement would be making these suggestions smarter.

Twixes committed:
Remove styled-components in favor of .scss files

Twixes committed:
Merge branch 'command-palette' of https://github.com/PostHog/posthog into command-palette

Twixes committed:
Remove redundant container class

Twixes committed:
Use insect (sic!) squeak instead of pig squeak

Twixes committed:
Show only unique palette results

mariusandra commented:
Hey, feedback on the UX feedback feedback:

> The input box is not cleared on palette close on purpose, this mimics Spotlight's behavior (which is pretty useful IMO).

In that case it would be great to "select all" the textfield as spotlight does.

> Definitely need to add some command suggestions for discoverability… Should be simple enough.

I think it's better with the suggestions!

> I don't quite understand "Clicking up/down shouldn't move the cursor", could you elaborate?

If you click up/down on the keyboard, the "caret" (not the cursor, sorry) in the textfield moves to the beginning of the line. 

![2020-10-12 14 01 17](https://user-images.githubusercontent.com/53387/95744338-6d651900-0c93-11eb-9908-d703a08e01a3.gif)

> Adding a calculator right now! 🔢

That's amazing! :D. I sent this as a bit of tongue-in-cheek comment... yet the result is great! Though I'd make a change and copy the result to clipboard on ENTER, instead of opening Wolfram Alpha. I'd assume it will be used for quick calculations that you then copy/past to slack, etc.

> Can hide this with a media query on narrow screens, though it shouldn't show up on mobile devices anyway (through a user agent check, this is because it's a slightly more reliable way to detect hardware keyboard-less devices).

I'd just make the box `width: calc(100% - 20px)` or something like that for smaller screens. Right now it's longer than the screen. All pages/widgets/etc should work for all resolutions :).

![2020-10-12 14 04 13](https://user-images.githubusercontent.com/53387/95744595-dc427200-0c93-11eb-886d-edfabef8c63a.gif)

> By "scroll", you mean show more than 5 results?

Nope, I mean if the user's screen height is low, I don't even see the 5 items.

![2020-10-12 14 05 37](https://user-images.githubusercontent.com/53387/95744711-10b62e00-0c94-11eb-96ea-e5b221504a84.gif)

Twixes committed:
Fix palette overlay

mariusandra commented:
I think I replied to the wrong thing with the feedback:

> Can hide this with a media query on narrow screens, though it shouldn't show up on mobile devices anyway (through a user agent check, this is because it's a slightly more reliable way to detect hardware keyboard-less devices).

1. Yep, please remove the CMD+K "button" on narrow screens via a media query
2. The "spotlight" should still work, but not be wider than the screen.

mariusandra commented on frontend/src/lib/components/CommandPalette/commandLogic.ts:
```diff
@@ -229,14 +235,21 @@ export const commandLogic = kea<commandLogicType<Command, CommandRegistrations>>
                     .map((result) => result.item)
-                    .sort((result) => (result.command.scope === GLOBAL_COMMAND_SCOPE ? 1 : -1))
+                const finalResults = guaranteedResults.concat(fusedResults)
+                return finalResults.sort((result) => (result.command.scope === GLOBAL_COMMAND_SCOPE ? 1 : -1))
```
Nit: this seems to work, though the `sort` function in JS takes two arguments. Right now with just comparing the first argument's scope with `"global"` is probably slightly less efficient than if you'd return 0 if the first and second argument's scopes are equal.

mariusandra commented on frontend/src/lib/components/CommandPalette/commandLogic.ts:
```diff
@@ -69,16 +74,17 @@ const GLOBAL_COMMAND_SCOPE = 'global'
+    const resultsWithCommand: CommandResult[] = results.map((result) => {
+        return { ...result, command }
+    })
+    resultsSoFar.push(...resultsWithCommand)
```
Functions that mutate their arguments are a good source of bugs. Since this is called inside a selector, which should be more or less following functional programming principles, I'd suggest to just return `resultsWithCommand` here and push them to `resultsSoFar` inside the selector.

mariusandra commented on frontend/src/lib/components/CommandPalette/commandLogic.ts:
```diff
@@ -4,6 +4,7 @@ import { commandLogicType } from 'types/lib/components/CommandPalette/commandLog
 import Fuse from 'fuse.js'
 import { dashboardsModel } from '~/models/dashboardsModel'
 import { Parser } from 'expr-eval'
+import _ from 'lodash'
```
Let's remove lodash from everywhere in the code. Most of what it provides can be done with ES6+ functions or small helpers in `lib/...`. Needing a [70kb](https://bundlephobia.com/result?p=lodash@4.17.20) (minified) dependency to get a few random elements from an array is not a good call.

mariusandra replied:
Closing since I mentioned it ~3x in the new batch of reviews

mariusandra commented on frontend/src/lib/components/CommandPalette/CommandResults.tsx:
```diff
@@ -10,7 +10,7 @@ interface CommandResultProps {
     handleSelection: (result: CommandResultType) => void
     focused?: boolean
-    setHoverResultIndex: Dispatch<SetStateAction<number | undefined | null>>
+    setHoverResultIndex: Dispatch<SetStateAction<number | undefined>>
```
why `undefined` and not `null`? 

In my mind `null` is more like "explicitly not there" and undefined is "not there by default". Plus if you ever store an `undefined` value in redux/kea, it'll throw an error, so it's wise to be consistent.

mariusandra commented on frontend/src/lib/components/CommandPalette/index.scss:
```diff
@@ -0,0 +1,96 @@
+    align-items: center;
+    width: 1rem;
+    height: 100%;
+}
```
Great for the BEM syntax, though since we're using SCSS, if possible it's nice to just nest elements in a very unique global class (`.command-palette { .container { .box { ... }  }  }`). This greatly reduces any kind of future conflicts... yet of course nested classes with simple names like `.palette { .icon { ... } }` will possibly have something bleed into them.

mariusandra reviewed (changes requested):
Hey! Spotted a few more things... plus the UX feedback from before.

I just checked the new commits and didn't go through the full list of [changed files](https://github.com/PostHog/posthog/pull/1819/files) in this PR. I'll do that once everything is again ready for review.

(Pro tip: before sending a PR for review, it's a good idea to just check all the modified files yourself as well... Even if everything seems to work fine. This is a good way to avoid sending `console.log` statements for review, etc)

Twixes committed:
Add powerful command building protocol CommandFlow

Twixes committed:
Fix minor issues

Twixes commented:
Here's what personal API key generation looks like now:

![hunter2](https://user-images.githubusercontent.com/4550621/95763964-1588db00-0cb0-11eb-8ecd-556acfe9f1ea.gif)

The cool thing about this is that this way command flows can be generated procedurally, offer many options at the same time, and be standarized just like regular commands at the same time (meaning definitions with a common interface instead of various hard-coded components). CC @paolodamico

paolodamico commented:
Replying to @Twixes,
> Here's what personal API key generation looks like now

Pretty cool that we now have a standardized way of creating custom commands! As a general comment I don't think it's bad to "hard-code" stuff some times, better to ship fast, test and iterate, than try to create the perfect product and never ship.

In UX terms, I think the second step (showing a create key command) is a bit confusing, as it can lead you to think you're using a different command instead of a step in the same command (think it might be even a bit more confusing on something like the feedback command where you have a large text field). Maybe just change the visuals so it doesn't look like another command?

Love that it takes you to the setup page (smooth scrolled) and shows the key there! 💪

Twixes commented:
Generally I agree that it's better to have something functioning shipped rather than something ideal in limbo, but the previous way doing more advanced commands would be a lot of ugly work, while with just some logic enhancement it's now a pretty trivial to use system.

Would the fact that it's a step be clearer if "Creating a Personal API Key" was shown constantly like so?

<img width="589" alt="Screen Shot 2020-10-12 at 18 22 45" src="https://user-images.githubusercontent.com/4550621/95769328-01e17280-0cb8-11eb-8838-aae44e35b834.png">

Twixes committed:
Always show scope when flow active

Twixes committed:
Use custom label icon

Twixes committed:
Add feedback sharing command

paolodamico committed:
Merge branch 'master' into command-palette

paolodamico committed:
block command input from being captured in screen recording

at least until we can figure out how to capture this info in a privacy-preserving way

Twixes committed:
Hide palette button on narrow screens

Twixes committed:
Merge branch 'command-palette' of https://github.com/PostHog/posthog into command-palette

Twixes committed:
Improve responsiveness

Twixes committed:
Fix palette feedback sending

Twixes committed:
Fix Esc handling

Twixes committed:
Add Message Sent info

Twixes committed:
Fix Message Sent info

Twixes committed:
Fix dashboard creation and null name handling

Twixes committed:
Rename Cy tests to JS convention

Twixes committed:
Add basic Cypress test

mariusandra commented on frontend/src/lib/components/CommandPalette/CommandResults.tsx:
```diff
@@ -0,0 +1,84 @@
+    const { activeResultIndex, commandSearchResults, commandSearchResultsGrouped } = useValues(commandPaletteLogic)
+    const { executeResult, onArrowUp, onArrowDown } = useActions(commandPaletteLogic)
+
+    useEventListener('keydown', (event: KeyboardEvent) => {
+        if (event.key === 'Enter' && commandSearchResults.length) {
+            const result = commandSearchResults[activeResultIndex]
+            const isExecutable = !!result.executor
+            if (isExecutable) executeResult(result)
+        } else if (event.key === 'ArrowDown') {
+            onArrowDown(commandSearchResults.length - 1)
+        } else if (event.key === 'ArrowUp') {
+            onArrowUp()
+        }
+    })
```
I would have also moved this into the logic (of course I would have...), though I guess it's fine here.

mariusandra commented on frontend/src/lib/components/CommandPalette/commandPaletteLogic.ts:
```diff
@@ -0,0 +1,681 @@
+// If CommandExecutor returns CommandFlow, flow will be entered
+export type CommandExecutor = () => CommandFlow | void
+
+export interface CommandResultTemplate {
+    icon: any // any, because Ant Design icons are some weird ForwardRefExoticComponent type
+    display: string
+    synonyms?: string[]
+    prefixApplied?: string
+    executor?: CommandExecutor | true // true means "just clear input"
+    guarantee?: boolean // show result always and first, regardless of fuzzy search
+}
+
+export type CommandResult = CommandResultTemplate & {
+    source: Command | CommandFlow
+    index?: number
+}
+
+export type CommandResolver = (
+    argument?: string,
+    prefixApplied?: string
+) => CommandResultTemplate[] | CommandResultTemplate | null
+
+export interface Command {
+    key: string // Unique command identification key
+    prefixes?: string[] // Command prefixes, e.g. "go to". Prefix-less case is dynamic base command (e.g. Dashboard)
+    resolver: CommandResolver | CommandResultTemplate[] | CommandResultTemplate // Resolver based on arguments (prefix excluded)
+    scope: string
+}
+
+export interface CommandFlow {
+    icon?: any
+    instruction?: string
+    resolver: CommandResolver | CommandResultTemplate[] | CommandResultTemplate
+    scope: string
+}
+
+export type CommandRegistrations = {
+    [commandKey: string]: Command
+}
+
```
nit: it's often a good idea to make a new `types.ts` file that all components can just import from. It helps keeps things a bit more organised...

Twixes replied:
That's true, though in this case I feel like these types are neat to have with the logic, as they are sort of the blueprints for it.

mariusandra replied:
yep, use your best judgement here :)

mariusandra commented on frontend/src/lib/components/CommandPalette/commandPaletteLogic.ts:
```diff
@@ -0,0 +1,681 @@
+                    return { ...commands, [command.key]: command }
+                },
+                deregisterCommand: (commands, { commandKey }) => {
+                    const cleanedCommands = { ...commands }
+                    delete cleanedCommands[commandKey]
+                    return cleanedCommands
```
```js
const { [commandKey]: discard, ...cleanedCommands } = commands // eslint-ignore-line
return cleanedCommands
```

mariusandra commented on frontend/src/lib/components/CommandPalette/commandPaletteLogic.ts:
```diff
@@ -0,0 +1,681 @@
+            if (values.isPaletteShown) window.posthog?.capture('palette shown')
+        },
+        executeResult: ({ result }: { result: CommandResult }) => {
+            if (result.executor === true) {
```
nit: `=== true` seems like a code smell, though I guess I understand the intent here. Probably nothing really to improve.

mariusandra commented on frontend/src/lib/components/CommandPalette/commandPaletteLogic.ts:
```diff
@@ -0,0 +1,681 @@
+            const { icon, index, ...cleanedResult } = result
+            // eslint-disable-next-line @typescript-eslint/no-unused-vars
+            const { resolver, ...cleanedCommand } = cleanedResult.source
+            cleanedResult.source = cleanedCommand
```
this gives a TS error

Twixes replied:
It does, but we're not strict about TS at the moment and I don't know what to do about this error (without adding a lot of boilerplate).

mariusandra replied:
Doesn't seem too complex?

![image](https://user-images.githubusercontent.com/53387/95864500-c0a79c00-0d65-11eb-8cc4-1dc3eb1eb5a0.png)

If you're not doing anything with this other than sending to posthog.capture, just cast it as a `Record<string, any>` and I think we're good to go.

We're not too strict about TS because of all the broken old/legacy code that doesn't work well with it... plus libraries that are a bit behind the times. I would love to enable stricter TS checks and definitely don't want to come back and revisit this issue then. I believe all new code should be good TS to the best of our abilities.

mariusandra commented on frontend/src/lib/components/CommandPalette/commandPaletteLogic.ts:
```diff
@@ -0,0 +1,681 @@
+        },
+        deregisterAllWithMatch: ({ keyPrefix }) => {
+            for (const command of Object.values(values.commandRegistrations)) {
+                if (command.key.includes(keyPrefix) || command.scope.includes(keyPrefix)) {
```
could `includes` eventually yield false positives? shouldn't it be `startsWith` instead? Otherwise I'd remove the variable to something other than key**Prefix**

Twixes replied:
It does. To be honest I'm not sure why this is the way it is, paging @EDsCODE?

EDsCODE replied:
ah there was something smelly with this. I meant this deregister function to just target the scope directly and nothing else but when I was doing the initial refactor I ran through this action too quickly so the operator here should actually just be a strict equality on command.scope. If someone pushes the change, I'd suggest making sure all the deregister calls are related to scope now and don't have a stray dependence on the key

Twixes replied:
Replaced `deregisterAllWithMatch` with `deregisterScope`.

mariusandra commented on frontend/src/lib/components/CommandPalette/commandPaletteLogic.ts:
```diff
@@ -0,0 +1,681 @@
+            }
+        },
+        setInput: async ({ input }, breakpoint) => {
+            await breakpoint(500)
```
nit: a 300ms debounce provides a smoother experience afaik 😅

mariusandra commented on frontend/src/lib/components/CommandPalette/commandPaletteLogic.ts:
```diff
@@ -0,0 +1,681 @@
+    return resultsWithCommand
+}
+
+export const commandPaletteLogic = kea<commandPaletteLogicType<Command, CommandRegistrations>>({
```
This should be changed to `commandPaletteLogicType<Command, CommandRegistrations, CommandResult, CommandFlow, RegExpCommandPairs>`

mariusandra commented on frontend/src/lib/components/CommandPalette/commandPaletteLogic.ts:
```diff
@@ -0,0 +1,681 @@
+                appUrlsLogic({ actionId: null }).selectors.appUrls,
+                appUrlsLogic({ actionId: null }).selectors.suggestions,
+            ],
+            (rawCommandRegistrations: CommandRegistrations, dashboards: DashboardType[]) => ({
```
this stupidly breaks typegen. Changing the line to

```js
(rawCommandRegistrations: CommandRegistrations, dashboards: DashboardType[]): CommandRegistrations => ({ ... })
```

... fixes it

mariusandra commented on frontend/src/lib/components/CommandPalette/commandPaletteLogic.ts:
```diff
@@ -0,0 +1,681 @@
+                    if (result.guarantee) guaranteedResults.push(result)
+                    else fusableResults.push(result)
+                }
+                fusableResults = _.uniqBy(fusableResults, 'display')
+                guaranteedResults = _.uniqBy(guaranteedResults, 'display')
```
could we replace lodash with some local `lib` utilities?

Twixes replied:
We probably could, but gotta say lodash is convenient…

mariusandra replied:
So is jQuery :D.

Trolling aside, I think in the worst case we should at least use the `lodash/uniqBy` type tree-shakeable imports, though having our own tiny helper functions is IMO the best solution. The less external dependencies the better.

mariusandra commented on frontend/src/lib/components/CommandPalette/commandPaletteLogic.ts:
```diff
@@ -0,0 +1,681 @@
+                          .search(argument)
+                          .slice(0, RESULTS_MAX)
+                          .map((result) => result.item)
+                    : _.sampleSize(fusableResults, RESULTS_MAX - guaranteedResults.length)
```
same lodash comment here

mariusandra commented on frontend/src/layout/TopContent/CommandPaletteButton.tsx:
```diff
@@ -0,0 +1,22 @@
+            title={isPaletteShown ? 'Hide Command Palette' : 'Show Command Palette'}
+        >
+            <SearchOutlined size={1} style={{ marginRight: '0.5rem' }} />
+            {platformCommandControlKey()} + K
```
This is quite a nit, but this looks better imo:
![image](https://user-images.githubusercontent.com/53387/95861891-622cee80-0d62-11eb-8a1a-423b92fb4d22.png)

mariusandra commented on frontend/src/lib/components/CommandPalette/index.tsx:
```diff
@@ -0,0 +1,52 @@
+            else if (input) setInput('')
+            // Lastly hide palette
+            else hidePalette()
+        } else if (event.key === 'k' && (event.ctrlKey || event.metaKey)) {
```
hmm.... so both CMD+K and CTRL+K work? shouldn't this also check the `isMac()` util?

Twixes replied:
Doesn't hurt to be universal I think

mariusandra commented on frontend/src/lib/utils.js:
```diff
@@ -450,3 +450,34 @@ export function copyToClipboard(value, description) {
     }
 }
+
+export function useOutsideClickHandler(refOrRefs, handleClickOutside) {
```
I'd put this under `lib/hooks`

mariusandra commented on frontend/src/scenes/insights/InsightTabs/TrendTab/trendCommandLogic.ts:
```diff
@@ -0,0 +1,11 @@
+import { commandPaletteLogicType } from 'types/lib/components/CommandPalette/commandPaletteLogicType'
+import { kea } from 'kea'
+
+export const trendCommandLogic = kea<commandPaletteLogicType<Command, CommandRegistrations>>({
+    connect: [commandPaletteLogic],
+    events: () => ({
+        afterMount: () => {},
+        beforeUnmount: () => {},
+    }),
+})
```
still not a useful logic :)

EDsCODE replied:
lost boilerplate....!

mariusandra commented on package.json:
```diff
@@ -41,12 +42,15 @@
         "kea-localstorage": "^1.0.2",
         "kea-router": "^0.4.0",
         "kea-window-values": "^0.0.1",
+        "lodash": "^4.17.20",
```
yup... I'd still remove this

mariusandra commented on posthog/api/test/test_person.py:
```diff
@@ -88,44 +97,128 @@ def test_filter_by_cohort(self):
+        another_org: Organization = Organization.objects.create()
+        another_team: Team = Team.objects.create(organization=self.organization)
+
+        another_person1: Person = Person.objects.create(
+            team=another_team, distinct_ids=["distinct_id", "x_another_one"],
+        )
+        another_person2: Person = Person.objects.create(
+            team=another_team, distinct_ids=["x_distinct_id_2"], properties={"email": "team2_another@gmail.com"},
+        )
```
Nit: this test would be more on point if the different persons shared the same distinct_id for whatever reason (e.g. "user_1" could be a common id)

Twixes replied:
I think it should be OK, but paging @paolodamico

paolodamico replied:
Actually this is already the case, `another_person1` shares a distinct ID ["distinct_id"] with `person` (which is a person from a different org) [line 157], to effectively test that even if a person in another team has the same ID as a person in your team, you won't see information from another team.

mariusandra reviewed (changes requested):
Another bunch of code comments (a lot of nits, feel free to ignore) and some UX points:

1. An annoying thing when coding on a mac: no scrollbars. I'm using a mouse, so scrollbars show by default (just like for all windows and most linux users) and here it's visible. Changing an `overflow: scroll` to `overflow: auto` somewhere will fix this.

![image](https://user-images.githubusercontent.com/53387/95860033-c306f780-0d5f-11eb-8729-ff7fc267860a.png)

2. Great for the scrollbar, but it doesn't scroll:

![2020-10-13 14 26 52](https://user-images.githubusercontent.com/53387/95860334-36106e00-0d60-11eb-9971-7976f3ae9258.gif)

3. At 320px some things get shortened to obscurity:

![image](https://user-images.githubusercontent.com/53387/95860207-02cddf00-0d60-11eb-9601-734b44cfdb5c.png)

Not sure if worth fixing now, but just letting items go on two lines could be useful.

Twixes committed:
Address feedback

Twixes commented:
Thanks for review @mariusandra, addressed all code-related points. As for small screen support, it doesn't seem worth the effort to make it perfect, as this feature is _really_ not intended for such devices at all, especially considering that there's no toggle button in such cases and a keyboard is necessary.

mariusandra commented:
@Twixes this seems very sudden... and I don't agree that all the points are resolved. I understand you were eager to get this merged, but I'm not sure it was the right call.

I still see a scrollbar even though there's nothing to scroll:

<img width="674" alt="Screenshot 2020-10-13 at 16 18 07" src="https://user-images.githubusercontent.com/53387/95872958-b7233180-0d6f-11eb-8379-f3a3fb89ec8d.png">

Please change a `overflow:scroll` to `overflow: auto` somewhere.

Also, this is definitely not resolved and I don't appreciate just closing it like that:

![image](https://user-images.githubusercontent.com/53387/95872233-dcfc0680-0d6e-11eb-9acd-c9bdc19c8a57.png)

mariusandra commented:
Regarding smaller screens, your call, though it also was not a nit. I'm fine to ignore it, but we can't assume what screens our users have and if we can't make features that are responsive, we shouldn't make them at all.

In this case, the solution would have just been to move the box to the top of the page and reduce the padding with a media query. You'll have 5 elements visible in no time.

![image](https://user-images.githubusercontent.com/53387/95873349-226d0380-0d70-11eb-9235-368156d4051f.png)

Twixes commented:
Oh, I apologize! This was quite sudden, sorry.
Definitely missed the `overflow: scroll` issue, fixed now in `master`.
As for small screens, this looks great even on an iPad Mini in its current form, any issue may arise only on smartphones, and IMHO it makes near-zero sense to make this superpolished on smartphones. As in, it will work anyway, just slightly less nicely, for the 0.01% of users operating their smartphone with an actual PC keyboard.
`lodash` I don't consider particularly harmful – ES got better over time, but it's still not there with a lot of utils and we're using `lodash` in other places in the codebase anyway. Though can remove the library, just in another PR dedicated to that (same amount of rewriting anyway, this PR I just wanted to ship for now).
